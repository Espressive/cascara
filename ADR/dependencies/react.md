# React

Our React version is tied to FE's version, we can bump only after they are doing bumping Ract. For now, all the following ecosystem members will be pinned to their current version:

- @types/react v16.8.0
- @types/react-dom v16.8.0
- react v16.13.1
- react-dom v16.13.1
- react-router v5.2.0
- react-router-dom v5.2.0
- react-scripts v4.0.3
- react-test-renderer v16.13.1
