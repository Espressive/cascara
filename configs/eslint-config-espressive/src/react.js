// "off" or 0 - turn the rule off
// "warn" or 1 - turn the rule on as a warning (doesn't affect exit code)
// "error" or 2 - turn the rule on as an error (exit code will be 1)

module.exports = {
  rules: {
    'react/boolean-prop-naming': 2,
    'react/default-props-match-prop-types': 2,
    'react/destructuring-assignment': 2,
    'react/jsx-fragments': 2,
    'react/jsx-no-duplicate-props': 2,
    'react/jsx-no-useless-fragment': 2,
    'react/jsx-pascal-case': 2,
    'react/jsx-sort-default-props': 2,
    'react/jsx-sort-props': 2,
    'react/no-children-prop': 2,
    'react/no-danger': 2,
    'react/no-find-dom-node': 2,
    'react/no-multi-comp': 2,
    'react/no-string-refs': 2,
    'react/no-this-in-sfc': 2,
    'react/no-typos': 2,
    'react/no-unused-prop-types': 2,
    'react/no-unused-state': 2,
    'react/prefer-stateless-function': 2,
    'react/self-closing-comp': 2,
    'react/sort-prop-types': 2,
    'react/style-prop-object': 2,
  },
};
