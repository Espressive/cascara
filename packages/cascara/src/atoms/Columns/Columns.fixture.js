import React from 'react';

import Columns from '.';

export default {
  withDefaultProps: (
    <>
      <p>
        This example shows the columns component without a `count` prop. That
        is, with default props.
      </p>
      <Columns>
        <div>Column 1</div>
        <div>Column 2</div>
      </Columns>
    </>
  ),
  withThreeColumns: (
    <>
      <p>
        This example shows the columns component with the <b>count</b> prop set
        to 3.
        <br />
        Please note that the number of children must be equal to the{' '}
        <b>count</b> prop, otherwise it will break the layout.
      </p>
      <Columns count={3}>
        <div>Column 1</div>
        <div>Column 2</div>
        <div>Column 3</div>
      </Columns>
    </>
  ),
};
