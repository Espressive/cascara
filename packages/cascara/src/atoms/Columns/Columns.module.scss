$masonry-gutter: 2rem;

._ {
  column-width: 28rem;
  column-gap: $masonry-gutter;

  & > * {
    -webkit-column-break-inside: avoid;
    page-break-inside: avoid;
    break-inside: avoid;
  }

  &.fixed {
    column-width: auto;
  }
}
