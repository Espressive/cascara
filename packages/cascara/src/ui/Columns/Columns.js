import React from 'react';
import pt from 'prop-types';
import { Role } from 'reakit';
import classNames from 'classnames/bind';

import Boundaries from '../../system-components/Boundaries';
import styles from './Columns.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /** HTML tag to allow polymorphism */
  as: pt.string,
  /** The section content */
  children: pt.oneOfType([pt.arrayOf(pt.node), pt.node]),
  /** Columns can have css class name */
  className: pt.string,
  /** The number of columns to allocate */
  count: pt.oneOfType([pt.number, pt.oneOf(['auto'])]),
  /** Columns can have inline styles */
  style: pt.shape(pt.object),
};

const Columns = ({
  as = 'div',
  count,
  children,
  className,
  style,
  ...rest
}) => {
  let mergedStyle = {
    columnCount: count,
  };

  if (typeof count === 'number' && style) {
    mergedStyle = {
      ...style,
      ...mergedStyle,
    };
  }

  return (
    <Boundaries>
      <Role
        className={cx('Columns', className, { fixed: Boolean(count) })}
        data-component='Columns'
        style={mergedStyle}
        {...rest}
      >
        {children}
      </Role>
    </Boundaries>
  );
};
Columns.propTypes = propTypes;

export { propTypes };
export default Columns;
