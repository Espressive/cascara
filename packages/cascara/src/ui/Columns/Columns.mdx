---
title: Columns
propTable: Columns.js
---

```jsx
---
title: An example use of Columns with default props (2 columns)
---
<Columns>
  <div key='Column 1'>Column 1</div>
  <div key='Column 2'>Column 2</div>
</Columns>
```

> The children count must match the `count` prop, otherwise the layout will be broken.
