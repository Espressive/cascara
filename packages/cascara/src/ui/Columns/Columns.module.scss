$masonry-gutter: 2em;
$widget-border-radius: 0.5em;
$widget-background: #f0f0f0;

.Columns {
  display: block;
  columns: 30rem;
  gap: $masonry-gutter;

  & > * {
    -webkit-column-break-inside: avoid;
    page-break-inside: avoid;
    break-inside: avoid;
  }
}
