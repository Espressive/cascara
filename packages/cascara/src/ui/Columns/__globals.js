import pt from 'prop-types';

export const PROP_TYPES = {
  /** The section content */
  children: pt.oneOfType([pt.arrayOf(pt.node), pt.node]),
  /** Columns can have css class name */
  className: pt.string,
  /** The number of columns to allocate */
  count: pt.number,
};
