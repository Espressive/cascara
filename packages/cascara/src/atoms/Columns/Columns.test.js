import React from 'react';
import { render } from '@testing-library/react';

import fixtures from './Columns.fixture';
import Columns from './Columns';

describe('<Columns />', () => {
  // run snapshot test for all fixtures.
  test('snapshot tests', async () => {
    for (const fixture in fixtures) {
      const view = render(fixtures[fixture]).container;

      expect(view).toMatchSnapshot();
    }
  });

  // this test expands the number of columns to 10.
  test('snapshot tests n', async () => {
    for (let columnsCount = 3; columnsCount < 11; columnsCount++) {
      const content = Array(columnsCount)
        .fill(1, 0)
        .map((_, idx) => (
          <div key={`Column ${idx + 1}`}>{`Column ${idx + 1}`}</div>
        ));

      const view = render(
        <Columns count={columnsCount}>{content}</Columns>
      ).container;

      expect(view).toMatchSnapshot();
    }
  });
});
