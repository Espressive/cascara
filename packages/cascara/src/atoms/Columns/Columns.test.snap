// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Columns /> snapshot tests 1`] = `
<div>
  <p>
    This example shows the columns component without a \`count\` prop. That is, with default props.
  </p>
  <div
    class="_"
    data-component="Columns"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests 2`] = `
<div>
  <p>
    This example shows the columns component with the 
    <b>
      count
    </b>
     prop set to 3.
    <br />
    Please note that the number of children must be equal to the
     
    <b>
      count
    </b>
     prop, otherwise it will break the layout.
  </p>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 3;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 1`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 3;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 2`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 4;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
    <div>
      Column 4
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 3`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 5;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
    <div>
      Column 4
    </div>
    <div>
      Column 5
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 4`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 6;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
    <div>
      Column 4
    </div>
    <div>
      Column 5
    </div>
    <div>
      Column 6
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 5`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 7;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
    <div>
      Column 4
    </div>
    <div>
      Column 5
    </div>
    <div>
      Column 6
    </div>
    <div>
      Column 7
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 6`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 8;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
    <div>
      Column 4
    </div>
    <div>
      Column 5
    </div>
    <div>
      Column 6
    </div>
    <div>
      Column 7
    </div>
    <div>
      Column 8
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 7`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 9;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
    <div>
      Column 4
    </div>
    <div>
      Column 5
    </div>
    <div>
      Column 6
    </div>
    <div>
      Column 7
    </div>
    <div>
      Column 8
    </div>
    <div>
      Column 9
    </div>
  </div>
</div>
`;

exports[`<Columns /> snapshot tests n 8`] = `
<div>
  <div
    class="_ fixed"
    data-component="Columns"
    style="column-count: 10;"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
    <div>
      Column 3
    </div>
    <div>
      Column 4
    </div>
    <div>
      Column 5
    </div>
    <div>
      Column 6
    </div>
    <div>
      Column 7
    </div>
    <div>
      Column 8
    </div>
    <div>
      Column 9
    </div>
    <div>
      Column 10
    </div>
  </div>
</div>
`;
